### DISCUSSION TITLE
Support time template

### DISCUSSION THREAD

**Discussion by Quorafind** [coddingtonbear/obsidian-web#1] [Ideas]
It will be useful when I need to add timestamp to the quick capture.

Like: {{time: HH:mm}} as 12:00

**Comment by coddingtonbear** [comment]
Definitely!  I think we're going to run into some challenges around what kind of templating is supported, though.  Most template libraries in the Javascript world rely on the `unsafe-eval` content security policy, and modern Chrome extensions aren't allowed to use that.  The particular library I've used here in this extension for supporting this doesn't use `unsafe-eval`, but it also doesn't support anything more than very basic templates (see: https://www.npmjs.com/package/micromustache).

There are a couple different options we could discuss if somebody wanted to try putting together a PR roughly in order of easiest to hardest:

1. Maybe we could just _not_ allow a time format to be specified?  It would be really trivial to provide a pre-formatted variable named something like `{{ timestamp }}` that stored a timestamp?
2. Maybe we could provide a handful of time formats at render time in an object so you could use template variables like `{{ timestamp.time.24h }}` or `{{ timestamp.time.12h }}`.  That would just be as easy as adding a template variable having content like:
```javascript
{
  ...
  timestamp: {
    time: {
      "24h": "time formatted in 24-h format",
      "12h": "time formatted in am/pm format",
    },
    date: ...
  }
```
3. Maybe we could [implement a sandbox](https://developer.chrome.com/docs/extensions/mv3/sandboxingEval/) for the template language so we could get access both to things like you've described in your request as well as also things like conditional blocks?  This is a bigger lift, but would open up a lot more avenues than the options discussed in (1) and (2) above.

**Comment by Quorafind** [comment]
Yes, someone loves `HH:mm:ss`, and someone loves `DD HH:mm` , the option 3 should be the best, but the option 2 is enough (but need work to specific HH mm ss and so on.

**Comment by coddingtonbear** [comment]
If it helps, this is where the context object used for providing you template variables comes from: https://github.com/coddingtonbear/obsidian-web/blob/master/src/popup.tsx#L83 .  If you were to post a PR, I'd recommend sticking to a very basic set of date and time options because it's likely that a sandboxed template environment will come at some point, and it'd be a shame to be stuck generating a bunch of rarely-used date/time formats forever just to prevent breaking compatibility with existing users' templates.  Maybe just:

For time:

* `24h`: 16:45:02
* `12h`: 4:45 PM

For date:

* `iso`: 2022-02-14
* `mdy`: 2/14/2022
* `dmy`: 14/2/2022

Or some other reasonable set :shrug: .

**Comment by coddingtonbear** [comment]
Good news -- I found a little time over the last few days to work on sandboxing the template renderer so we can do fancier things than what the current template renderer supports.  Through that, I've added a handlebars helper named `date` that you'll be able to use like:  `{{date "yyyy-MM-dd HH:mm:ss"}}` once I've polished it up a little.

You can find these changes in the `sandbox` branch if you're curious to have a look before they're released.

**Comment by coddingtonbear** [comment]
& just submitted v2.0.0 to the Chrome Web Store for review with this feature (and a few others).  Have a look at https://github.com/coddingtonbear/obsidian-web/releases/tag/2.0.0 for details about what's included in that release, but the particular feature you probably care about most is the addition of the `{{date}}` template helper.  Cheers!

**Reply by Quorafind** [reply]
Nice!
